File: backend/app/routers/addresses.py
"""
Router de direcciones del cliente.
"""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.address import AddressCreate, AddressUpdate, AddressResponse
from app.core.dependencies import get_current_user
from app.core.exceptions import NotFoundException, ForbiddenException
from app.models.user import User
from app.models.address import Address

router = APIRouter(prefix="/addresses", tags=["Direcciones"])


@router.get("", response_model=list[AddressResponse])
def list_addresses(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Listar direcciones del usuario."""
    return db.query(Address).filter(Address.user_id == current_user.id).all()


@router.post("", response_model=AddressResponse)
def create_address(
    data: AddressCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Agregar nueva dirección."""
    if data.is_default:
        db.query(Address).filter(
            Address.user_id == current_user.id
        ).update({"is_default": False})

    address = Address(user_id=current_user.id, **data.model_dump())
    db.add(address)
    db.commit()
    db.refresh(address)
    return address


@router.put("/{address_id}", response_model=AddressResponse)
def update_address(
    address_id: int,
    data: AddressUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Actualizar dirección."""
    address = db.query(Address).filter(Address.id == address_id).first()
    if not address:
        raise NotFoundException("Dirección no encontrada")
    if address.user_id != current_user.id:
        raise ForbiddenException()

    update_data = data.model_dump(exclude_unset=True)
    if update_data.get("is_default"):
        db.query(Address).filter(
            Address.user_id == current_user.id
        ).update({"is_default": False})

    for key, value in update_data.items():
        setattr(address, key, value)
    db.commit()
    db.refresh(address)
    return address


@router.delete("/{address_id}")
def delete_address(
    address_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Eliminar dirección."""
    address = db.query(Address).filter(Address.id == address_id).first()
    if not address:
        raise NotFoundException("Dirección no encontrada")
    if address.user_id != current_user.id:
        raise ForbiddenException()
    db.delete(address)
    db.commit()
    return {"message": "Dirección eliminada"}

File: frontend/src/services/addresses.ts
import { api } from "@/lib/api";

export interface Address {
  id: number;
  user_id: number;
  label: string;
  full_address: string;
  reference?: string | null;
  district?: string | null;
  city?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  is_default: boolean;
  created_at: string;
}

export interface AddressInput {
  label: string;
  full_address: string;
  reference?: string;
  district?: string;
  city?: string;
  latitude?: number;
  longitude?: number;
  is_default?: boolean;
}

export async function listAddresses() {
  const { data } = await api.get<Address[]>("/addresses");
  return data;
}

export async function createAddress(payload: AddressInput) {
  const { data } = await api.post<Address>("/addresses", payload);
  return data;
}

export async function updateAddress(id: number, payload: Partial<AddressInput>) {
  const { data } = await api.put<Address>(`/addresses/${id}`, payload);
  return data;
}

export async function deleteAddress(id: number) {
  const { data } = await api.delete<{ message: string }>(`/addresses/${id}`);
  return data;
}
